### PR TITLE
Implementation of the full BMCA

### PR DESCRIPTION
**Pull request by diondokter**
This probably needs a good review.
It is easy to miss a rule here.

### CODE REVIEW

File: src/bmc/bmca.rs [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+        announce_message: &AnnounceMessage,
+        current_time: Timestamp,
+    ) {
+        self.foreign_master_list
+            .register_announce_message(announce_message, current_time);
+    }
+
+    /// Takes the Erbest from this port. If called before two announce intervals have passed since the last call
+    /// the returned value will like to be None. So don't do that.
+    pub fn take_best_port_announce_message(
```
Comment (davidv1992) [inline src/bmc/bmca.rs:51]: What is the purpose of this function? Per subclause b of 9.3.2.3 of the standard, I would expect to have something that removes those considered but not selected as E_rbest from the foreign master list, but not something that removes all or erbest itself from the list.

That would also make more sense in light of the requirement to run the BMCA at least once per announce interval

File: src/bmc/foreign_master.rs [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        current_time: Timestamp,
+        announce_interval: TimeInterval,
+    ) {
+        self.purge_old_messages(current_time, announce_interval);
+        self.announce_messages
+            .push((announce_message, current_time));
+    }
+
+    /// Calculates Erbest and returns its announce message
+    pub fn get_best_announce_message(&self) -> (AnnounceMessage, Timestamp) {
```
Comment (davidv1992) [inline src/bmc/foreign_master.rs:70]: What does this function do given that this currently is only a single foreign master? Based on 9.3.2.5 subclause b, it seems to me that it should just return the most recent message, but it seems to do more than that

File: src/bmc/foreign_master.rs [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+            own_port_identity,
+        }
+    }
+
+    /// Returns all of the qualified announce messages.
+    /// All foreign masters will be removed
+    pub fn take_qualified_announce_messages(
+        &mut self,
+        current_time: Timestamp,
+    ) -> impl Iterator<Item = ForeignMaster> + '_ {
```
Comment (davidv1992) [inline src/bmc/foreign_master.rs:115]: See above comment on the bmca implementation: when is this useful?

File: src/bmc/foreign_master.rs [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+            // No, insert a new foreign master
+            self.foreign_masters.push(ForeignMaster::new(
+                *announce_message,
+                current_time,
+                self.own_port_identity,
+            ));
+        }
+    }
+
+    fn find_foreign_master_mut(
```
Comment (davidv1992) [inline src/bmc/foreign_master.rs:171]: I think per the rust naming convention this and the next function should be get_...., but that is a minor issue

**Review by davidv1992 — COMMENTED** [review]
I have some doubts whether this implements proper behaviour for the Foreign Master List, which should be cleared before we can merge this

**Comment by diondokter** [comment]
I think this is ready now

**Pull request merged**